Add SanitizeFieldNames setting (#549)
elastic/apm-agent-dotnet: docs/configuration.asciidoc

| true    | Boolean 
|============

[float]
[[config-sanitize-field-names]]
==== `SanitizeFieldNames` (added[1.2])

> Sometimes it is necessary to sanitize the data sent to Elastic APM, e.g. remove sensitive data.

Reviewer: e.g. --> for example (**egg** sample)
i.e. --> in other words (**i**n **e**ssence)

Is removing sensitive data an example of sanitizing data? Or is it the same thing? If it's the same thing, which I think it is, this should be an i.e. instead.
```suggestion
Sometimes it is necessary to sanitize, i.e., remove, sensitive data sent to Elastic APM.
```

Author: Yeah, agreed, it's the same.

This config contains a list of wildcard patterns of field names which should be sanitized. These apply for example to HTTP headers and `application/x-www-form-urlencoded` data.

This option supports the wildcard `*`, which matches zero or more characters.
Examples: `/foo/*/bar/*/baz*`, `*foo*`.
Matching is case insensitive by default.

Prepending an element with `(?-i)` makes the matching case sensitive.

NOTE: Data in the query string is considered non-sensitive,
as sensitive information should not be sent in the query string.

> See https://www.owasp.org/index.php/Information_exposure_through_query_strings_in_url for more information

Reviewer: ```suggestion
See https://www.owasp.org/index.php/Information_exposure_through_query_strings_in_url[owasp.org] for more information.
```

NOTE: Review the data captured by Elastic APM carefully to make sure it does not capture sensitive information.
If you do find sensitive data in the Elasticsearch index,

> you should add an additional entry to this list (make sure to also include the default entries).

Reviewer: ```suggestion
you should add to the default entries on this list.
```

Author: Hmm, this is interesting, if you don’t mind I’d discuss this a little bit.

So if you don’t configure anything, we have a default list of keys and based on those we’ll sanitize the HTTP headers and the body. If you still see some sensitive data, then you should use the setting. 

But the thing is, as soon as you use this setting the defaults will go away. So not configuring `SanitizeFieldNames` gives you a default list (list is above in this doc), but then adding `”SanitizeFieldNames”: “foo”` will only have `”foo`”, the defaults are gone - we'll only sanitize the `foo` HTTP header and form body values with `foo`. So you can’t really “add to the defaults”. 

I think having the `(make sure to also include the default entries)` makes this clear. Does this makes sense?

Btw. all this is copied from Java (or maybe the Go doc is mixed into it at some point) - so none of this text is really written by me.

Author: > Is this typical, i.e., will a developer expect this? Now I'm concerned that the docs don't really explain that the default goes away if you configure SanitizeFieldNames.

That is a very good point. I'm not sure that everyone expects this behaviour. We can add some more text to explain this, I think it'd make the docs better.

Reviewer: 👍 Working on it

> you should add an additional entry to this list (make sure to also include the default entries).

Reviewer: Can I get a little wild? What about something like this?

```asciidoc
Sometimes it is necessary to sanitize, i.e., remove, sensitive data sent to Elastic APM.
This config accepts a list of wildcard patterns of field names which should be sanitized.
These apply for example to HTTP headers and `application/x-www-form-urlencoded` data.

The wildcard, `*`, matches zero or more characters, and matching is case insensitive by default.
Prepending an element with `(?-i)` makes the matching case sensitive.
Examples: `/foo/*/bar/*/baz*`, `*foo*`.

You should review the data captured by Elastic APM carefully to make sure it does not contain sensitive information.
If you do find sensitive data in your Elasticsearch index,
you should add an additional entry to this list.
Make sure to include the default entries as well, as setting a value here will overwrite the defaults. 

NOTE: Data in the query string is considered non-sensitive,
as sensitive information should not be sent in the query string.
See https://www.owasp.org/index.php/Information_exposure_through_query_strings_in_url[owasp.org] for more information.
```

You don't have to use all of this. The important changes are in paragraph three. I removed the admonition paragraph (NOTE) and clarified that the defaults get overwritten.

Author: I like it, it captures all the details nicely. I'm fine with taking this. 👍

Author: I like it, it captures all the details nicely. I'm fine with taking this. 👍

Author: Pushed it.
